src/Hint/Monad.hs

```diff
@@ -110,6 +134,52 @@ monadHint _ _ d = concatMap (f Nothing Nothing) $ childrenBi d
         isHsDo (L _ HsDo{}) = True
         isHsDo _ = False
 
+gratuitouslyMonadic :: LHsDecl GhcPs -> [Idea]
+gratuitouslyMonadic e@(L _ d) = case d of
+  ValD _ func@(FunBind _ (L _ n) (MG _ (L _ ms))) -> do
+    guard $ fname /= "main"  -- Account for "main = pure ()" test
+    guard $ all gratuitouslyMonadicExpr $ allMatchExprs ms
+    pure $ rawIdea
+      Suggestion
+      "Unnecessarily monadic"
+      (locA $ getLoc e)
+      (unsafePrettyPrint e)
+      (Just $ unwords ["Demote", "`" <> fname <> "`", "to a pure function"])
+      []
+      []
+    where
+      fname = occNameString $ rdrNameOcc n
+      -- Iterate over all of the patterns of the function, as well as all of the guards
+      allMatchExprs ms = [expr | L _ (Match _ _ _ (GRHSs _ xs _)) <- ms, L _ (GRHS _ _ expr) <- xs]
+  _ -> []
+
+-- | Handles expressions of both these forms:
+--     pure x
+--     pure $ f x
+--
+-- Also recurses into `do` blocks to check whether it consists entirely
+-- (excluding any Let bindings) of "Body Statements" with
+-- such expressions. This catches at least some real-world
+-- sightings of the phenomenon.
+gratuitouslyMonadicExpr :: LHsExpr GhcPs -> Bool
+gratuitouslyMonadicExpr x =
+  case simplifyExp x of
+    L _ (HsApp _ (L _ (HsVar _ (L _ myFunc))) _) ->
+      occNameString (rdrNameOcc myFunc) `elem` ["pure", "return"]
+    L _ (HsDo _ _ (L _ statements)) -> all isGratuitouslyMonadicBodyStatement $
+      filter (not . isLetStatement) statements
+    _ -> False
+  where
+    isGratuitouslyMonadicBodyStatement statement = case statement of
+      (L _ (BodyStmt _ x _ _)) -> gratuitouslyMonadicExpr x
+      _ -> False
+    
+    isLetStatement statement = case statement of
+      (L _ (LetStmt _ _)) -> True
+      _ -> False
```

Comment:
we should probably lift this function into ghc-lib-parser-ex and reference it from there.